### PR TITLE
♻️ REFACTOR: Directive plugins

### PR DESCRIPTION
**Pull request by chrisjsewell**
Note for development I have put the code in the `src/directives2` folder, but eventually this will replace `src/directives`. The admonition plugin is functionally complete (~~except option conversion/validation~~) and now I will start looking at the figure and math ones.

So this PR essentially re-writes the whole of the directive plugin code/logic.
This is necessary since, among other things, for the current code:

1. It doesn't handle parsing options
2. It doesn't handle nested admonitions (in fact the parsing completely expects if you have any)
3. It doesn't parse admonition titles as Markdown
4. There is no obvious way to map from Sphinx <-> these plugins

The new logic is designed to closely mirror sphinx/myst-parser and follows as such:

1. The `src/directives/base.ts:pluginDirectiveBase` is defined. 

All specific directive plugins will rely on this being loaded first.
It simply walks through the token stream and converts all code fences, whose token.info matches the directive regex (i.e. looks like `{name} argument`), to a `directive_base` type token, changes the token.info to just the name and stores the argument in token.meta.arg. See `tests/directiveBase.spec.ts` for an example.

Having this as a separate plugin separates out the logic of identifying a directive, from how it is processed/rendered.
It means that it is possible, in the future, to change our directive syntax without affecting any of the other directive plugins. Or even people can use different plugins to generate these `directive_base` tokens.

2. `src/directives/parseStructure.ts:parseStructure` is defined.

This is a function converts the first line and body of a `directive_base` token to the structure of a directive: args (list of strings), options (dict), body (string), dependent on the particular directive specification (e.g. how many arguments it requires and whether it has a body). The code is directly adapted from `myst_parser/parse_directives.py`.

This code is generic to all directives, and will be called in the individual plugins.

3. The `src/directives2/admonition.ts:pluginDirectiveAdmonition` is defined.

This acts on `directive_base` tokens of specific names (admonition, note, etc); first calling `parseStructure`, performing a nested (block level only) parse on the body string of the directive, then replacing the original token with the following list of tokens:

- `open_admonition` (tag = aside class = 'admonition' + type of admonition + any set in the `class` option of the directive)
- `open_admonition_title` (tag = div, class = 'admonition-title')
- `inline` (containing the title text)
- `close_admonition_title`
- `open_admonition_body`(tag = div, class = 'admonition-body')
- list of body tokens output from the nested parsing
- `close_admonition_body`
- `close_admonition`

Outputting this "structure" of tokens means that you don't actually require a specific render method for admonitions, since the default markdown-it renderer will deal with it. Where possible this how plugins should be implemented; generating a token structure that directly mirrors the HTML structure and requires no/minimal custom rendering.

See for `tests/directiveAdmonition.spec.ts` for more examples, but this is what the parse looks like:

`````
````{admonition} arg string
:class: class1 class2

content
```{admonition} arg string2
content2
```
````
`````

is parsed to tokens:

```javascript
[
'{"type":"open_admonition","tag":"aside","attrs":[["class","admonition' +
    '"]],"map":[0,6],"nesting":1,"level":0,"children":null,"content":"","markup":"","info":"","meta":null,"block":false,"hidden":false}',
'{"type":"open_admonition_title","tag":"div","attrs":[["class","admonition-title"]],"map":null,"nesting":1,"level":0,"children":null,"content":"","markup":"","info":"","meta":null,"block":false,"hidden":false}',
'{"type":"inline","tag":"","attrs":null,"map":[0,0],"nesting":0,"level":0,"children":[{"type":"text","tag":"","attrs":null,"map":null,"nesting":0,"level":0,"children":null,"content":"arg string","markup":"","info":"","meta":null,"block":false,"hidden":false}],"content":"arg string","markup":"","info":"","meta":null,"block":false,"hidden":false}',
'{"type":"close_admonition_title","tag":"div","attrs":null,"map":null,"nesting":-1,"level":0,"children":null,"content":"","markup":"","info":"","meta":null,"block":false,"hidden":false}',
'{"type":"open_admonition_body","tag":"div","attrs":[["class","admonition-body"]],"map":[1,5],"nesting":1,"level":0,"children":null,"content":"","markup":"","info":"","meta":null,"block":false,"hidden":false}',
'{"type":"paragraph_open","tag":"p","attrs":null,"map":[1,2],"nesting":1,"level":0,"children":null,"content":"","markup":"","info":"","meta":null,"block":true,"hidden":false}',
'{"type":"inline","tag":"","attrs":null,"map":[1,2],"nesting":0,"level":1,"children":[{"type":"text","tag":"","attrs":null,"map":null,"nesting":0,"level":0,"children":null,"content":"content","markup":"","info":"","meta":null,"block":false,"hidden":false}],"content":"content","markup":"","info":"","meta":null,"block":true,"hidden":false}',
'{"type":"paragraph_close","tag":"p","attrs":null,"map":null,"nesting":-1,"level":0,"children":null,"content":"","markup":"","info":"","meta":null,"block":true,"hidden":false}',
'{"type":"open_admonition","tag":"aside","attrs":[["class","admonition' +
    '"]],"map":[2,5],"nesting":1,"level":0,"children":null,"content":"","markup":"","info":"","meta":null,"block":false,"hidden":false}',
'{"type":"open_admonition_title","tag":"div","attrs":[["class","admonition-title"]],"map":null,"nesting":1,"level":0,"children":null,"content":"","markup":"","info":"","meta":null,"block":false,"hidden":false}',
'{"type":"inline","tag":"","attrs":null,"map":[2,2],"nesting":0,"level":0,"children":[{"type":"text","tag":"","attrs":null,"map":null,"nesting":0,"level":0,"children":null,"content":"arg string2","markup":"","info":"","meta":null,"block":false,"hidden":false}],"content":"arg string2","markup":"","info":"","meta":null,"block":false,"hidden":false}',
'{"type":"close_admonition_title","tag":"div","attrs":null,"map":null,"nesting":-1,"level":0,"children":null,"content":"","markup":"","info":"","meta":null,"block":false,"hidden":false}',
'{"type":"open_admonition_body","tag":"div","attrs":[["class","admonition-body"]],"map":[3,4],"nesting":1,"level":0,"children":null,"content":"","markup":"","info":"","meta":null,"block":false,"hidden":false}',
'{"type":"paragraph_open","tag":"p","attrs":null,"map":[3,4],"nesting":1,"level":0,"children":null,"content":"","markup":"","info":"","meta":null,"block":true,"hidden":false}',
'{"type":"inline","tag":"","attrs":null,"map":[3,4],"nesting":0,"level":1,"children":[{"type":"text","tag":"","attrs":null,"map":null,"nesting":0,"level":0,"children":null,"content":"content2","markup":"","info":"","meta":null,"block":false,"hidden":false}],"content":"content2","markup":"","info":"","meta":null,"block":true,"hidden":false}',
'{"type":"paragraph_close","tag":"p","attrs":null,"map":null,"nesting":-1,"level":0,"children":null,"content":"","markup":"","info":"","meta":null,"block":true,"hidden":false}',
'{"type":"close_admonition_body","tag":"div","attrs":null,"map":null,"nesting":1,"level":0,"children":null,"content":"","markup":"","info":"","meta":null,"block":false,"hidden":false}',
'{"type":"close_admonition","tag":"aside","attrs":null,"map":null,"nesting":-1,"level":0,"children":null,"content":"","markup":"","info":"","meta":null,"block":false,"hidden":false}',
'{"type":"close_admonition_body","tag":"div","attrs":null,"map":null,"nesting":1,"level":0,"children":null,"content":"","markup":"","info":"","meta":null,"block":false,"hidden":false}',
'{"type":"close_admonition","tag":"aside","attrs":null,"map":null,"nesting":-1,"level":0,"children":null,"content":"","markup":"","info":"","meta":null,"block":false,"hidden":false}'
]
```

which is rendered to:

```html
<aside class="admonition class1 class2"><div class="admonition-title">arg string</div><div class="admonition-body"><p>content</p>
<aside class="admonition"><div class="admonition-title">arg string2</div><div class="admonition-body"><p>content2</p>
<div></aside><div></aside>
```

### CODE REVIEW

**Comment by chrisjsewell** [comment]
cc'ing @choldgraf @rowanc1 and @agoose77 for an initial look

**Comment by chrisjsewell** [comment]
I have now also made the HTML tags for open_admonition, open_admonition_title and open_admonition_body a setting of the plugin. The use case for this, is that you can set `main` as `details` and `title` as `summary`, and voila you've got collapsible admonitions 🎉 
See e.g. https://github.com/executablebooks/sphinx-panels/blob/master/sphinx_panels/scss/panels/_dropdown.scss, how you can style these as an accordion

**Review by rowanc1 — COMMENTED** [review]
Standardizing on the tokens coming out makes sense. As this comes in I will need to update downstream packages to compensate for these changes.

It looks like this is removing the reliance on `markdown-it-container`, which may have been giving some problems with nesting of directives. Once this replaces, we should remove that library from dependencies.

This will also require some changes to the numbering (figure/equation/references) that I have done so far. Thinking that some of that may not be wanted/needed by consumers if they are only interested in the structure of the outputs (at the AST level).

Additionally the docs will need to change:
https://executablebooks.github.io/markdown-it-myst/overview.html#creating-a-new-directive

On a broader level, are you planning other large structural changes? Can we discuss those in an issue if you are?

**Comment by chrisjsewell** [comment]
Thanks for the comments 👍, here some responses:

> Standardizing on the tokens coming out makes sense. As this comes in I will need to update downstream packages to compensate for these changes.

Yes indeed, and that will be covered by semantic versioning. If you haven't already, I would suggest pinning to 0.0.x

> It looks like this is removing the reliance on markdown-it-container, which may have been giving some problems with nesting of directives. Once this replaces, we should remove that library from dependencies.

Indeed, this was removed a long time ago in the Python code

> This will also require some changes to the numbering (figure/equation/references) that I have done so far. Thinking that some of that may not be wanted/needed by consumers if they are only interested in the structure of the outputs (at the AST level).

As mentioned in my initial comment, I will be rewriting all of them

> Additionally the docs will need to change:
executablebooks.github.io/markdown-it-myst/overview.html#creating-a-new-directive

👍 

> On a broader level, are you planning other large structural changes? Can we discuss those in an issue if you are?

As mentioned I will be rewriting all the directives. I haven't had a look at the roles yet, but quite possibly aspects of that as well. Again the key point is to have them aligned with the sphinx/myst-parser code (plus make best use of markdown-it). If you haven't already I would strongly suggest you look through the myst-parser code and also sphinx, particularly: https://www.sphinx-doc.org/en/master/extdev/appapi.html, it will be necessary to understand that before understanding why certain decisions have been taken here

File: src/directives2/parseStructure.ts [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+function parseDirectiveArguments(
+  firstLine: string,
+  fullSpec: IDirectiveSpec
+): string[] {
+  let args = firstLine.trim() ? firstLine.trim()?.split(/\s+/) : []
+  const totalArgs =
+    (fullSpec.required_arguments || 0) + (fullSpec.optional_arguments || 0)
+  if (args.length < (fullSpec.required_arguments || 0)) {
+    throw new DirectiveParsingError(
```
[the remark this answers is not in the document]
Comment (chrisjsewell) [reply src/directives2/parseStructure.ts:193]: yes they are logged as warnings, the parsing succeeds but the directive is not rendered. there is already a TODO in the admonition plugin, where the error is caught and the token discarded (presently without warning)

File: tests/directiveAdmonition.spec.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  '````{admonition} arg string\ncontent\n```{admonition} arg string2\ncontent2\n```\n````'
+
+describe('base directive plugin', () => {
+  const mdit = MarkdownIt().use(pluginDirectiveBase).use(pluginDirectiveAdmonition)
+  describe('creates admonition token when', () => {
+    it('parses a matched admonition fence', () => {
+      const tokens = mdit.parse(basicDirective, {})
+      // console.log(tokens.map(token => JSON.stringify(token)))
+      expect(tokens.map(token => JSON.stringify(token))).toEqual([
+        '{"type":"open_admonition","tag":"aside","attrs":[["class","admonition' +
```
[the remark this answers is not in the document]
Comment (chrisjsewell) [reply tests/directiveAdmonition.spec.ts:19]: yes indeed; the directives/roles in the first instance should be building up the token stream (i.e. AST), this allows for other plugins to potentially modify it "later".

File: src/directives2/admonition.ts [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+      if (token.info !== 'admonition') {
+        classes = ['admonition', token.info]
+      }
+      if (options.class && options.class.length) {
+        classes.push(...options.class)
+      }
+      adToken.attrSet('class', classes.join(' '))
+      newTokens.push(adToken)
+      const adTokenTitle = new Token('open_admonition_title', tags.title, 1)
+      adTokenTitle.attrSet('class', 'admonition-title')
```
[the remark this answers is not in the document]
Comment (chrisjsewell) [reply src/directives2/admonition.ts:168]: as mentioned there is no special renderer, its all handled by markdown-it's default one

File: src/directives2/parseStructure.ts [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+  option_spec?: {
+    [key: string]: OptionSpecConverter
+  }
+}
+
+export const DirectiveSpecDefaults: IDirectiveSpec = {
+  required_arguments: 0,
+  optional_arguments: 0,
+  final_argument_whitespace: false,
+  has_content: false,
```
[the remark this answers is not in the document]
Comment (chrisjsewell) [reply src/directives2/parseStructure.ts:40]: correct

File: src/directives2/base.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+ *
+ * @param regex to identify a fence info line as a directive
+ * @param logError console.error if the render method is called
+ */
+export default function pluginDirectiveBase(
+  md: MarkdownIt,
+  regex = DIRECTIVE_PATTERN,
+  logError = true
+): void {
+  // TODO also convert colon-fences
```
[the remark this answers is not in the document]
Comment (chrisjsewell) [reply src/directives2/base.ts:59]: Its a separate plugin: https://github.com/executablebooks/mdit-py-plugins/blob/master/mdit_py_plugins/colon_fence.py

File: src/directives2/admonition.ts [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+  },
+  warning: {
+    spec: baseAdmonitionSpec,
+    title: 'Warning'
+  }
+}
+
+const defaultTags = {
+  main: 'aside',
+  title: 'div',
```
[the remark this answers is not in the document]
Comment (chrisjsewell) [reply src/directives2/admonition.ts:92]: yep I can change this to `header`; I was having a look if there was any good tag for the body, but didn't see one? I'm not convinced though thats these can fully replace class names, for specificity

**Comment by chrisjsewell** [comment]
@rowanc1 so obviously this PR has essentially been superseded by https://github.com/executablebooks/markdown-it-docutils.
But, as discussed, I've updated the PR with the stuff I was playing around with for a "full" sphinx-like implementation; principally in `src/appComponents.ts`, and `src/app.ts`, to have a record of that.
and also `docs/design.md` should eventually have a home somewhere

**Comment by rowanc1** [comment]
Closing this! See #34.

**Pull request closed without merging**